// appium-windows-driver/build/lib/commands/general.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _teen_process = require("teen_process");

var _logger = _interopRequireDefault(require("../logger"));

const commands = {};
exports.commands = commands;

commands.getWindowSize = async function getWindowSize() {
  const size = await this.winAppDriver.sendCommand('/window/size', 'GET');

  if (_lodash.default.isPlainObject(size)) {
    return size;
  }

  _logger.default.info('Cannot retrieve window size from WinAppDriver. ' + 'Falling back to Windows Forms to calculate dimensions');

  return await getScreenSize();
};

async function getScreenSize() {
  const {
    stdout
  } = await (0, _teen_process.exec)('powershell', ['-command', 'Add-Type -AssemblyName System.Windows.Forms;[System.Windows.Forms.Screen]::PrimaryScreen.Bounds.Size']);

  _logger.default.debug(`Screen size information retrieved: ${stdout}`);

  const match = /^\s*(True|False)\s+(\d+)\s+(\d+)/m.exec(stdout);

  if (!match) {
    throw new Error('Cannot retrieve the screen size. Check the server log for more details');
  }

  return {
    width: parseInt(match[2], 10),
    height: parseInt(match[3], 10)
  };
}

commands.getScreenshot = async function getScreenshot() {
  return await this.winAppDriver.sendCommand('/screenshot', 'GET');
};

var _default = commands;
exports.default = _default;require('source-map-support').install();
